tests/test/opaque_types.rs

```diff
@@ -46,6 +46,49 @@ fn opaque_reveal() {
     }
 }
 
+#[test]
+fn opaque_where_clause() {
+    test! {
+        program {
+            struct Ty { }
+
+            trait Clone { }
+            trait Trait { }
+            impl Trait for Ty { }
+            opaque type T: Clone where T: Trait = Ty;
+
+            struct Vec<U> { }
+            impl<U> Trait for Vec<U> { }
+
+            opaque type S<U>: Clone where U: Trait = Vec<U>;
+        }
+
+        goal {
+            if (T: Trait) {
+                WellFormed(T)
+            }
+        } yields {
+            "Unique; substitution []"
+        }
+
+        goal {
+            WellFormed(T)
+        } yields {
+            "No possible solution"
+        }
+
+        goal {
```

Comment:
maybe we should add a test `forall<U> { WellFormed(S<U>) }`? which would error, presumably.